bigquery/google/cloud/bigquery/_pandas_helpers.py

```diff
@@ -176,5 +207,163 @@ def to_parquet(dataframe, bq_schema, filepath):
+    if "$" in table.table_id:
+        raise ValueError(
+            "Reading from a specific partition is not currently supported."
+        )
+    if "@" in table.table_id:
+        raise ValueError("Reading from a specific snapshot is not currently supported.")
+
+    read_options = bigquery_storage_v1beta1.types.TableReadOptions()
+    if selected_fields is not None:
+        for field in selected_fields:
+            read_options.selected_fields.append(field.name)
+
+    requested_streams = 0
+    if preserve_order:
+        requested_streams = 1
+
+    session = bqstorage_client.create_read_session(
+        table.to_bqstorage(),
+        "projects/{}".format(project_id),
+        read_options=read_options,
+        requested_streams=requested_streams,
+    )
+
+    # We need to parse the schema manually so that we can rearrange the
+    # columns.
+    schema = json.loads(session.avro_schema.schema)
+    column_names = [field["name"] for field in schema["fields"]]
+
+    # Avoid reading rows from an empty table. pandas.concat will fail on an
+    # empty list.
+    if not session.streams:
+        yield pandas.DataFrame(columns=column_names)
+        return
+
+    total_streams = len(session.streams)
+
+    # Use _DownloadState to notify worker threads when to quit.
+    # See: https://stackoverflow.com/a/29237343/101923
+    download_state = _DownloadState()
+
+    # Create a queue to collect frames as they are created in each thread.
+    worker_queue = queue.Queue()
+
+    with concurrent.futures.ThreadPoolExecutor(max_workers=total_streams) as pool:
+        try:
+            # Manually submit jobs and wait for download to complete rather
+            # than using pool.map because pool.map continues running in the
+            # background even if there is an exception on the main thread.
+            # See: https://github.com/googleapis/google-cloud-python/pull/7698
+            not_done = [
+                pool.submit(
+                    _download_dataframe_bqstorage_stream,
+                    download_state,
+                    bqstorage_client,
+                    column_names,
+                    dtypes,
+                    session,
+                    stream,
+                    worker_queue,
+                )
+                for stream in session.streams
+            ]
+
+            while not_done:
+                done, not_done = _nowait(not_done)
+                for future in done:
+                    # Call result() on any finished threads to raise any
+                    # exceptions encountered.
+                    future.result()
+
+                try:
+                    frame = worker_queue.get(timeout=_PROGRESS_INTERVAL)
+                    yield frame
+                except queue.Empty:  # pragma: NO COVER
+                    continue
+
+            # Return any remaining values after the workers finished.
+            while not worker_queue.empty():
```

Comment:
Just trying to understand the possible scenarios, in what cases could there be leftovers in the queue that we would have to clean up here?

If a `_download_dataframe_bqstorage_stream()` future completes and ends up in the `done` list, it means that one of the following has happened when we reach this code:
- A frame was put in the queue and the function returned normally. In this case we would have already taken the frame from the queue and yielded it a few lines above.
Unless ...
- ...an exception occurred in one of the `done` workers. The `future.result()` would propagate that exception, we would jump to the `finally` block, and this queue-clearing loop would not be executed.

I might be missing something, thus mostly bringing this up as a sanity check.

Reply:
On the very last iteration of the "wait for all worker threads to be done" loop, it's possible that we pulled everything from the queue but between when we `yield`ed the last item and we checked for the workers to be done a worker wrote to the queue.

* reader: get
* reader: yield
* (function using these results does stuff)
* worker: put
* worker: done
* reader: all done!
* reader: oh wait, there's still something on the queue